src/v/raft/replicate_entries_stm.cc

```diff
@@ -40,37 +40,29 @@ ss::future<append_entries_request> replicate_entries_stm::share_request() {
           });
     });
 }
-ss::future<result<append_entries_reply>> replicate_entries_stm::do_dispatch_one(
-  vnode n,
-  append_entries_request req,
-  ss::lw_shared_ptr<std::vector<ss::semaphore_units<>>> units) {
-    using ret_t = result<append_entries_reply>;
-
-    if (n == _ptr->_self) {
-        auto f = _ptr->flush_log()
-                   .then([this, units]() {
-                       auto lstats = _ptr->_log.offsets();
-                       auto last_idx = lstats.committed_offset;
-                       append_entries_reply reply;
-                       reply.node_id = _ptr->_self;
-                       reply.target_node_id = _ptr->_self;
-                       reply.group = _ptr->group();
-                       reply.term = _ptr->term();
-                       // we just flushed offsets are the same
-                       reply.last_dirty_log_index = last_idx;
-                       reply.last_committed_log_index = last_idx;
-                       reply.result = append_entries_reply::status::success;
-                       return ret_t(std::move(reply));
-                   })
-                   .handle_exception(
-                     []([[maybe_unused]] const std::exception_ptr& ex) {
-                         return ret_t(errc::leader_flush_failed);
-                     });
 
-        _dispatch_sem.signal();
-        return f;
-    }
-    return send_append_entries_request(n, std::move(req));
+ss::future<result<append_entries_reply>> replicate_entries_stm::flush_log() {
+    using ret_t = result<append_entries_reply>;
+    auto f = _ptr->flush_log()
+               .then([this]() {
+                   auto new_committed_offset = _dirty_offset;
```

Comment:
I will add the comment in here, the `_dirty_offset` is sampled after successful append to leader disk. It is done like this to be able to append to leader log before  before receiving the previous append entries responses. In this particular case it is enough to use this sampled `_dirty_offset` since in `replicate_entries_stm` we do not wait for any future offsets.

Reply:
no need to change the comment, but the missing piece in the new comment is that the member variable _dirty_offset is read and written under the consensus semaphore and that is what makes it safe.